File: coding interview guide/栈和队列/设计一个有getMin功能的栈.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.data_stack = list()
        self.min_stack = list()

    def push(self, x: int) -> None:
        if len(self.min_stack) == 0:
            self.min_stack.insert(0, x)
        elif x <= self.getMin():
            self.min_stack.insert(0, x)

        self.data_stack.insert(0, x)

    def pop(self) -> None:
        if (len(self.data_stack) == 0):
            assert RuntimeError("your stack is empty")
        num = self.data_stack[0]
        del self.data_stack[0]
        if (num == self.getMin()):
            del self.min_stack[0]

    def top(self) -> int:
        if (len(self.data_stack) == 0):
            assert RuntimeError("your stack is empty")
        else:
            return self.data_stack[0]

    def getMin(self) -> int:
        if (len(self.min_stack) == 0):
            assert RuntimeError("your stack is empty")
        else:
            return self.min_stack[0]

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: coding interview guide/栈和队列/设计一个有getMin功能的栈.cpp
#include <iostream>
#include <stack>
#include <exception>
using namespace std;

class MinStack{
public:
	void push(int newNum){
		if(min_stack.empty()){
			min_stack.push(newNum);
		}else if(newNum < getMin()){
			min_stack.push(newNum);
		}else{
			min_stack.push(getMin());
		}
		data_stack.push(newNum);
	}
	int pop(){
		if(data_stack.empty()){
			throw runtime_error("your stack is empty");
		}
		int num = data_stack.top();
		data_stack.pop();
		min_stack.pop();
		return num;
	}
	int getMin(){
		if(min_stack.empty()){
			throw runtime_error("your stack is empty");
		}
		return min_stack.top();
	}
	int top(){
		if(data_stack.empty()){
			throw runtime_error("your stack is empty");
		}
		return data_stack.top();
	}
private:
	stack<int> data_stack;
	stack<int> min_stack;
};

